azurerm/internal/services/mssql/mssql_server_transparent_data_encryption_resource.go

```diff
@@ -0,0 +1,244 @@
+	defer cancel()
+
+	log.Printf("[INFO] preparing arguments for MSSQL Encrypted Protector creation.")
+
+	serverId, err := parse.ServerID(d.Get("server_id").(string))
+
+	if err != nil {
+		return err
+	}
+
+	// Normally we would check if this is a new resource, but the way encryption protector works, it always overwrites
+	// whatever is there anyways. Compounding the issue is that SQL Server creates an instance of encryption protector
+	// which causes the isNewResource check to fail because we are trying to create the encryption as a separate resource
+	// and encryption protector is already present. The reason we create encryption protector as a separate resource is
+	// because after the SQL server is created, we need to grant it permissions to AKV, so encryption protector can use those
+	// keys are part of setting up TDE
+
+	var serverKey sql.ServerKey
+
+	// Default values for Service Managed keys. Will update to AKV values if key_vault_key_id references a key.
+	serverKeyName := ""
+	serverKeyType := sql.ServiceManaged
+
+	keyVaultKeyId := strings.TrimSpace(d.Get("key_vault_key_id").(string))
+
+	// If it has content, then we assume it's a key vault key id
+	if keyVaultKeyId != "" {
+		// Update the server key type to AKV
+		serverKeyType = sql.AzureKeyVault
+
+		// Set the SQL Server Key properties
+		serverKeyProperties := sql.ServerKeyProperties{
+			ServerKeyType: serverKeyType,
+			URI:           &keyVaultKeyId,
+		}
+		serverKey.ServerKeyProperties = &serverKeyProperties
+
+		// Set the encryption protector properties
+		keyId, err := keyVaultParser.ParseNestedItemID(keyVaultKeyId)
+
+		if err != nil {
+			return fmt.Errorf("Unable to parse key: %q: %+v", keyVaultKeyId, err)
+		}
+
+		// Make sure it's a key, if not, throw an error
+		if keyId.NestedItemType == "keys" {
+			keyName := keyId.Name
+			keyVersion := keyId.Version
+
+			// Extract the vault name from the keyvault base url
+			idURL, err := url.ParseRequestURI(keyId.KeyVaultBaseUrl)
+
+			if err != nil {
+				return fmt.Errorf("Unable to parse key vault hostname: %s", keyId.KeyVaultBaseUrl)
+			}
+
+			hostParts := strings.Split(idURL.Host, ".")
+			vaultName := hostParts[0]
+
+			// Create the key path for the Encryption Protector. Format is: {vaultname}_{key}_{key_version}
+			serverKeyName = fmt.Sprintf("%s_%s_%s", vaultName, keyName, keyVersion)
+		} else {
+			return fmt.Errorf("Key vault key id must be a reference to a key, but got: %s", keyId.NestedItemType)
+		}
+	}
+
+	// Service managed doesn't require a key name
+	encryptionProtectorProperties := sql.EncryptionProtectorProperties{
+		ServerKeyType: serverKeyType,
+		ServerKeyName: &serverKeyName,
+	}
+
+	// Only create a server key if the properties have been set
+	if serverKey.ServerKeyProperties != nil {
+		// Create a key on the server
+		futureServers, err := serverKeysClient.CreateOrUpdate(ctx, serverId.ResourceGroup, serverId.Name, serverKeyName, serverKey)
+		if err != nil {
+			return err
```

Comment:
We should wrap this error

```suggestion
			return fmt.Errorf("creating/updating server key for %s: %+v", serverId, err)
```